### PR TITLE
Implement `Serialize` without going through JsonObject

### PR DESCRIPTION
**Pull request by zyla**
As described in <https://github.com/georust/geojson/issues/152>, serialization by constructing a JsonObject has to essentially clone all data, which requires extra allocations.

After this change, we pass data directly to the `Serializer`.

`JsonObject` and `JsonValue` conversion impls were rewritten to delegate to `Serialize`. This requires some unfortunate `panic`s where we expect `Serialize` to produce a JSON object, but I think it's better than duplicating the serialization logic.

Benchmark results (compared vs `4cfbf3a`):

```
serialize geojson::FeatureCollection struct (countries.geojson)
                        time:   [990.01 µs 1.0025 ms 1.0172 ms]
                        change: [-74.789% -74.267% -73.763%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

serialize custom struct (countries.geojson)
                        time:   [2.1757 ms 2.2237 ms 2.2718 ms]
                        change: [-40.176% -38.891% -37.671%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

serialize custom struct to geo-types (countries.geojson)
                        time:   [2.3386 ms 2.3570 ms 2.3776 ms]
                        change: [-42.699% -41.782% -40.960%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  7 (7.00%) high mild
  3 (3.00%) high severe
```

Custom structs still have to go through JsonObject, so there isn't as much improvement there.

Expected JSON outputs in tests had changed, because now we produce JSON object keys in order of the `serialize_entry` calls, instead of `serde_json::Map` order.

- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/master/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

### CODE REVIEW

**Comment by urschrei** [comment]
Thanks for the contribution @zyla! Could you say more and / or provide an example (ideally in the form of a test) of the kind of data that can trigger a panic using this PR?

**Comment by lnicola** [comment]
I don't think this can panic. It assumes that serializing a `JsonObject` (which is a `serde_json::Map`) always results in an object, which seems like a reasonable assumption.

File: src/feature.rs [hunk 1]
```diff
@@ -182,7 +159,26 @@ impl Serialize for Feature {
     where
         S: Serializer,
     {
-        JsonObject::from(self).serialize(serializer)
+        let mut map = serializer.serialize_map(None)?;
```
Comment (lnicola) [inline src/feature.rs:162]: I wanted to suggest passing in the length here, but `serde_json` doesn't seem to use it.

**Review by michaelkirk — CHANGES_REQUESTED** [review]
This is great, thank you!

I agree with @lnicola that the panics "should never happen" in this case.

Since `serde_json::to_value` will attempt to serialize any input, the API needs to be fallible, but ours never will because all our fields are serializable given that they are themselves JsonObjects, None (which is serializable), or other primitive types (String, Vec<f64>, etc.).

It's probably worth documenting this assumption though. It took me a minute to reason through it.

File: src/feature.rs [hunk 2]
```diff
@@ -182,7 +159,26 @@ impl Serialize for Feature {
...
+        map.serialize_entry("type", "Feature")?;
+        map.serialize_entry("geometry", &self.geometry)?;
+        if let Some(ref properties) = self.properties {
+            map.serialize_entry("properties", properties)?;
+        } else {
+            map.serialize_entry("properties", &JsonObject::new())?;
```
[the remark this answers is not in the document]
Comment (zyla) [reply src/feature.rs]: Sure. Changed the behavior and added a test.

**Comment by zyla** [comment]
> It's probably worth documenting this assumption though. It took me a minute to reason through it.

Thanks for the review. Added some comments about that.

**Review by michaelkirk — APPROVED** [review]

**Pull request merged**